### PR TITLE
Add mode 'contain' to image resizer

### PR DESCRIPTION
**Pull request by yapsr**
This mode 'contain' will resize the original image to a new image with the exact given width and height respecting the original aspect ratio.

### CODE REVIEW

**Comment by yapsr** [comment]
Related to #307.

**Comment by yapsr** [comment]
These are the result of all the modes, including the new 'contain' mode:
image

File: src/Database/Attach/Resizer.php [hunk 1]
```diff
@@ -1,15 +1,15 @@
 <?php namespace October\Rain\Database\Attach;
 
+use Exception;
 use File as FileHelper;
 use Symfony\Component\HttpFoundation\File\File as FileObj;
-use Exception;
```
Comment (LukeTowers) [inline src/Database/Attach/Resizer.php]: Thanks :)

File: src/Database/Attach/Resizer.php [hunk 2]
```diff
@@ -231,34 +199,47 @@ protected function getHeight()
     }
 
     /**
-     * Receives the original but rotated image
-     * according to exif orientation
-     * @return resource (gd)
+     * Sets resizer options. Available options are:
+     *  - mode: Either exact, portrait, landscape, auto or crop.
```
Comment (LukeTowers) [inline src/Database/Attach/Resizer.php]: missing contain documented here

File: src/Database/Attach/Resizer.php [hunk 3]
```diff
@@ -275,36 +256,47 @@ public function resize($newWidth, $newHeight, $options = [])
         /*
          * Sanitize input
          */
-        $newWidth = (int) $newWidth;
-        $newHeight = (int) $newHeight;
+        $newWidth = (int)$newWidth;
+        $newHeight = (int)$newHeight;
```
Comment (LukeTowers) [inline src/Database/Attach/Resizer.php]: Please leave the spaces here

File: src/Database/Attach/Resizer.php [hunk 4]
```diff
@@ -275,36 +256,47 @@ public function resize($newWidth, $newHeight, $options = [])
...
+        } else {
+            $imageResized = imagecreatetruecolor($optimalWidth, $optimalHeight);
+        }
 
         // Retain transparency for PNG and GIF files
-        imagecolortransparent($imageResized, imagecolorallocatealpha($imageResized, 0, 0, 0, 127));
+        $transparency = imagecolorallocatealpha($imageResized, 0, 0, 0, 127);
+
+        // fill empty image with transparent color, only required for contain mode
+        imagefill($imageResized, 0, 0, $transparency);
```
Comment (LukeTowers) [inline src/Database/Attach/Resizer.php]: If it's only required for the contain mode shouldn't be wrapped in an if statement to limit it to just that mode?

File: src/Database/Attach/Resizer.php [hunk 5]
```diff
@@ -275,36 +256,47 @@ public function resize($newWidth, $newHeight, $options = [])
...
+
+        imagecolortransparent($imageResized, $transparency);
         imagealphablending($imageResized, false);
         imagesavealpha($imageResized, true);
 
         // get the rotated the original image according to exif orientation
         $rotatedOriginal = $this->getRotatedOriginal();
 
+        // get offset to center images for mode 'contain'
+        list($offsetX, $offsetY) = $this->getOffset($optimalWidth, $optimalHeight, $newWidth, $newHeight);
```
Comment (LukeTowers) [inline src/Database/Attach/Resizer.php]: Does this only apply to the contain mode as well? If so, should it be in it's own conditional?

**Comment by yapsr** [comment]
@LukeTowers Thank you for reviewing the code. I will process your requested changes asap.

**Comment by LukeTowers** [comment]
@yapsr thanks!

**Comment by yapsr** [comment]
@LukeTowers I have processed your requested changes in this commit:
[https://github.com/octobercms/library/commit/7e574cbcb1e3347e0af7edff84f48bc509daa20c](url)

Could you please look into it?  I have tested it locally and it seems to work fine.

**Comment by LukeTowers** [comment]
@yapsr Sorry for not realizing this sooner but this is essentially the same functionality as #307 right? If so, what's the difference between this PR and that one?

**Comment by yapsr** [comment]
This PR adds the mode "contain", which returns an image with exactly the dimensions as given, containing the original image in a transparent PNG.
For example, if I would want to resize a square image to a new image of 400x300 pixels, I get a transparent image of 400x300 pixels with the square image in the middle.

I have run the mode "fit" in PR #307 but as far as I can see it returns the exact same result as with the "auto" method. The image returned has different dimensions than given.

This mode "contain" differs from the mode "exact" in that it respects the dimension ratio. The mode "exact" stretches the original image.

**Comment by LukeTowers** [comment]
@yapsr do you still want this merged?

**Comment by w20k** [comment]
@LukeTowers I’d love to get this  merged. One of the requirements for an upcoming project.

**Comment by yapsr** [comment]
> @yapsr do you still want this merged?

I do, but I haven't found the time yet to write the required tests.

**Comment by LukeTowers** [comment]
Perhaps @Notuom can help?

**Comment by w20k** [comment]
@yapsr shouldn't you also add `contain` here:
```
protected function getDimensions($newWidth, $newHeight)
    {
        $mode = $this->getOption('mode');

        switch ($mode) {
```

**Comment by w20k** [comment]
Closing as it has been over a month since any activity on this occurred and we are trying to figure out what issues are still relevant. If this is still something that you would like to see through to fruition please respond and we can get the ball rolling.

_Note: Better to revise it based on the `fit` option and make a new PR, if it's still something you'd like to add._

**Pull request closed without merging**